[ownership] Extract out from SILOwnershipVerifier the OperandOwnershi…

gottesmm opened the pull request:
…pKindMapClassifier

rdar://44667493

----

This is to get some initial feedback. It doesn't pass all tests yet. I think borrow checking is still broken as well. I am most interested in getting feedback about the OperandOwnershipKindMap API, etc.

gottesmm committed:
[ownership] Extract out from SILOwnershipVerifier the OperandOwnershipKindMapClassifier

NOTE: This is not the final form of how operand ownership restraints will be
represented. This patch is instead an incremental change that extracts out this
functionality from the ownership verifier as a pure refactor.

rdar://44667493

gottesmm commented:
I am probably going to split off parts of this CPU by slowing slicing parts off. I want to get something working first though.

rjmccall commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -64,26 +64,26 @@ struct ErrorBehaviorKind {
   /// A cache of dead-end basic blocks that we use to determine if we can
   /// ignore "leaks".
-  DeadEndBlocks &DEBlocks;
+  DeadEndBlocks &deadEndBlocks;
```
We're generally favoring this convention for locals now, I think, but I don't think we've actually agreed to use it for fields.

gottesmm replied:
I thought that was the style that we were using in SILGen, no? I thought we were trying to standardize on SILGen style.

Maybe we should have a discussion about this on swift-dev and get something written up in a small style guide so there is an easy reference for the forgetful (i.e. me) among us. ; ).

gottesmm replied:
I just created a thread for this on swift-dev: https://forums.swift.org/t/style-used-in-sil-siloptimizer/16495. I don't want this to be held up by that conversation. I am happy to go back in afterwards and fix the casing if needed.

rjmccall reviewed (approved):
Seems reasonable, although I'm not really an expert in your algorithm here.

atrick commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -7913,6 +7924,15 @@ class FullApplySite : public ApplySite {
+  }
+
+  bool isIndirectResultOperand(const Operand &oper) const {
+    return oper.getOperandNumber() < getNumIndirectSILResults();
```
This isn't right is it? Shouldn't it be `getCalleeArgIndex(oper) < getNumIndirectSILResults`?

gottesmm replied:
As we discussed via chat:

```
bool isIndirectResultOperand(const Operand &oper) const {
  return getCalleeArgIndex(oper) < getSILArgIndexOfFirstParam();
}
```

gottesmm commented on include/swift/SIL/SILValue.h:
```diff
@@ -55,6 +55,33 @@ static inline llvm::hash_code hash_value(ValueKind K) {
+  /// has left the scope of the borrowed SSA value and said value can not be
+  /// used.
+  MustBeInvalidated,
+};
```
I don't think we can pass in general trivial values to RC related operations. That is something we should discuss. That being said, I do think IsConsuming_t is a better name for this.

gottesmm commented on include/swift/SIL/SILValue.h:
```diff
@@ -147,6 +169,23 @@ struct ValueOwnershipKind {
+  ///
+  /// This is MustBeInvalidated for Owned and MustBeLive for all other ownership
+  /// kinds.
+  UseLifetimeConstraint getDefaultForwardingLifetimeConstraint() const {
```
I am fine with that.

gottesmm commented on include/swift/SIL/SILValue.h:
```diff
@@ -343,6 +382,148 @@ class SILValue {
+/// A map from a ValueOwnershipKind that an operand can accept to a
+/// UseLifetimeConstraint that describes the effect that the operand's use has
+/// on the underlying value. If a ValueOwnershipKind is not in this map then
+/// matching an operand with the value results in an ill formed program.
```
Andy and I spoke about this. I agree with him that this is the direction in which things need to move. I want to make those (related) change in a different PR since I am trying to keep this PR as close to a refactor as possible.

atrick replied:
Cool. Thanks @gottesmm .

gottesmm commented:
Back on this.

gottesmm committed:
[semantic-arc-opts] Change Semantic ARC Opts to run only on the stdlib so it only sees ownership verified SIL.

The reason why I am doing this is now that once we have the
OperandOwnershipKindMap I can write this optimization in a more robust,
aggressive manner. My hope is that I can eliminate all copy_value of guaranteed
parameters all of whose uses could accept a guaranteed parameter. This is given
to me by the OperandOwnershipKindMap.

Additionally, I found that the way the analysis was using the OwnershipVerifier
was not sound on non-ownership verified SIL. Rather than fix it, I just turned
it off for that case.

rdar://44667493

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci smoke benchmark

gottesmm commented:
@swift-ci test compiler performance

gottesmm commented:
@swift-ci test compiler performance

gottesmm commented:
@swift-ci smoke benchmark
